### PR TITLE
new protocol, overloads to list.sort, sorted; prep. to fix #4051

### LINKED ISSUE #4051: Should Mypy warn about potential invalid arguments to `max`?
I was surprised that Mypy didn't flag an issue where I was calling `max` with an int and a None, which throws a TypeError at runtime.

For example, in the following code, should I not expect Mypy to warn that `top` is potentially None, which wouldn't work with `max` ?

```python
from typing import Optional
top: Optional[int] = None
print(max(5, top))
```

### PR DESCRIPTION
**Pull request by ramalho**
As agreed with @JelleZijlstra, this PR introduces the `_SupportsLessThan` protocol and a type variable bounded to it to check arguments to `list.sort` and `sorted` on Python 3 (on Python 2, these checks do not reflect reality: everything supports `__lt__`, even `None` and `object`).

When this PR is merged, I will send another one using `_SupportsLessThan` to check arguments to `min` and `max`, fixing #4051 with lots of overloads.

### CODE REVIEW

File: stdlib/2and3/builtins.pyi [hunk 1]
```diff
@@ -943,7 +947,10 @@ class list(MutableSequence[_T], Generic[_T]):
     def remove(self, __value: _T) -> None: ...
     def reverse(self) -> None: ...
     if sys.version_info >= (3,):
-        def sort(self, *, key: Optional[Callable[[_T], Any]] = ..., reverse: bool = ...) -> None: ...
+        @overload
+        def sort(self: List[_LT], *, key: None = ..., reverse: bool = ...) -> None: ...
+        @overload
+        def sort(self, *, key: Callable[[_T], _LT] = ..., reverse: bool = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/builtins.pyi:953]: You should remove the `= ...` from this overload (same with `sorted`).

Comment (ramalho) [reply stdlib/2and3/builtins.pyi:953]: I am sorry, which `= ...` should I remove?

Comment (JelleZijlstra) [reply stdlib/2and3/builtins.pyi:953]: Sorry, the one for `key`. In this overload the `key` argument is required.

Comment (ramalho) [reply stdlib/2and3/builtins.pyi]: Yes, it took me a few minutes to understand what you meant. Thanks!

I did not touch the Python 2 version of `sort`, but I was surprised to see that neither `cmp` nor `key` are `Optional`... I wonder how that's working now.

**Comment by ramalho** [comment]
There was something wrong with `stubtest py39`, and @hauntsaninja apparently fixed the problem by removing Python 3.9 from the CI but now I don't know how to run the checks again... Sorry.

**Pull request closed without merging**